### DISCUSSION TITLE
Help with bin file

### DISCUSSION THREAD

**Discussion by CyBaart** [arendst/Tasmota#19439] [Support request]
Can someone please help. I want to use logical expressions in my Tasmota rules but this requires me to compile my own bin file which I meticulously followed instructions and did, enabling logic before compiling of course. But when I try to update my ESP32C3, it bricks and I have to restore the original one and all is well except I cant use IF THEN statements.. If anyone has a compiled bin file with logic enabled and knows it works, can you please put it somewhere I can download eg dropbox etc.

Thanks guys.

**Comment by barbudor** [comment]
How did you compile your own binary ? With which tools ? TasmoCompiler ? VisualCode ?  Gitpod ?
How did you proceed ?
Normally to compile `tasmota32c3` with the following lines in `user_config_override.h` or in the dedicated box for TasmoCompiler
```
#define USE_EXPRESSION
#define SUPPORT_IF_STATEMENT
```

Note that the above additions applies to is between the DO and the ENDON of the rule.
Not to part between the ON and the DO
So is that really what you mean when writing "logical expressions in my Tasmota rules"

Also on an ESP32C3 you have the Berry language that is much more powerful than Rules to write complex conditions, decisions and expressions.

**Reply by sfromis** [reply]
With ESP32, you do **not** need to do a build to enable expressions or `IF` in rules, as these features are **already** included in default builds. But yeah, the Berry language is a great enhancement over rules.

**Reply by CyBaart** [reply]
Thanks for your replies.

I used Gitpod to compile it. So you are saying that the rule is something like this;
ON _trigger_ DO _action_ IF _condition_  ENDIF ENDON
Am I correct?

Basically I want to turn a heater on and off to raise the hot tub water temperature but only if the ambient is below a certain temperature. Would that be simpler in Berry?

**Reply by sfromis** [reply]
The general syntax is:
`ON trigger DO IF (condition) action ENDIF ENDON`
And with ESP32, you need no custom build to have this work.

*Very* simple things are not "simpler in Berry", but needing a few more code lines. If/when you start to improve it, doing it with rules starts to get pretty arcane to get around the "quite different" ways rules work, as it is not similar to a programming language. One "funny" thing steepening the rules learning curve is that when you do not do it **exactly** right, you get no helpful messages, only things not working. Every little misunderstanding is punished hard, and debugging is not simple.

**Comment by barbudor** [comment]
> ON trigger DO action IF condition ENDIF ENDON

No, most like :
```
ON trigger DO IF (condition) command; command; command ELSE command; command; command ENDIF
```
Like it is written in the docs: https://tasmota.github.io/docs/Rules/#conditional-rules

> Basically I want to turn a heater on and off to raise the hot tub water temperature but only if the ambient is below a certain temperature. 

There is most chances that you don't even need IF for that
Assuming for exemple your sensor message is like:
```
SENSOR = {"Time":"2021-01-02T09:09:44","DS18B20-1":{"Id":"00000566CC39","Temperature":25.3},"DS18B20-2":{"Id":"0000059352D4","Temperature":48.2},"TempUnit":"C"}
```
with the 1st DS18x20 being the ambiant temperature and the 2nd one is the hot tube.
And with `mem1` being the ambiant temperature threshold and `mem2` the hot tube threshold
Something like that would probably do the job:
```
Rule1
  ON tele-DS18B20-1<%mem1% DO var1 1 ENDON
  ON tele-DS18B20-1>%mem1% DO Backlog var1 0; Power1 0 ENDON
  ON tele-DS18B20-2>%mem2% DO Power1 0 ENDON
  ON tele-DS18B20-2<%mem2% DO Power1 %var1% ENDON
Rule1 1
Rule1 5
```

**Reply by CyBaart** [reply]
Thanks for all replies. I think I have enough to build with. Cheers.
